src/testing/TestFields.tsx

```diff
@@ -0,0 +1,151 @@
+import React, { FC } from 'react';
+import { Box, Text, Label, Input, Checkbox, Textarea, Radio } from 'theme-ui';
+import { FormikContextType, FormikValues } from 'formik';
+import { Message } from 'retranslate';
+import * as yup from 'yup';
+
+import { TestType, ObjectSchema, FieldSchema } from '../api';
+
+interface TestFieldsProps {
+  form: FormikContextType<FormikValues>;
+  testType: TestType;
+}
+
+export const TestFields: FC<TestFieldsProps> = ({ form, testType }) => {
+  const fieldError = (key: keyof TestType['resultsSchema']['properties']) =>
+    form.touched[key] && form.errors[key] ? <Text>{form.errors[key]}</Text> : null;
+
+  return (
+    <>
+      {Object.entries(testType.resultsSchema.properties).map(([key, value]) => {
```

Comment:
I might have created a JsonSchemaForm component, that just takes a schema, a (nullable) initial model, and fires an onChange when the model changes.

The TW one also takes a 'submitted' prop, which triggers validation on any fields that haven't already done it.

Reply:
Yup, we considered this as a next refactor — will look into it, but probably postpone it till after the launch.